File: src/main/java/net/serebryansky/common/service/impl/CityServiceImpl.java
package net.serebryansky.common.service.impl;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.serebryansky.city.model.City;
import net.serebryansky.common.service.CityService;
import org.slf4j.MDC;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import static net.serebryansky.common.util.LoggingUtil.logOnNext;

@Slf4j
@RequiredArgsConstructor
public class CityServiceImpl implements CityService {

    private final WebClient cityClient;

    @Override
    public Mono<City> getCity(String cityId) {
        // log.info("Get city {}", cityId);
        return Mono.just(cityId)
                .doOnEach(logOnNext(p -> log.info("Get city {}", p)))
                .then(Mono.deferContextual(contextView -> {
                            String requestId = contextView.get("CONTEXT_KEY");
                            String invoker = contextView.get("INVOKER");
                            return cityClient.get()
                                    .uri("/cities/{id}", cityId)
                                    .header("X-Request-ID", requestId)
                                    .header("X-Source", invoker)
                                    .retrieve()
                                    .bodyToMono(City.class);
                        }
                ));
    }
}

File: src/main/kotlin/net/serebryansky/common/service/impl/CityServiceImpl.kt
package net.serebryansky.common.service.impl

import kotlinx.coroutines.reactive.awaitFirst
import kotlinx.coroutines.reactor.mono
import mu.KotlinLogging
import net.serebryansky.city.model.City
import net.serebryansky.common.requestKey
import net.serebryansky.common.service.CityService
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import reactor.core.publisher.Mono.*

class CityServiceImpl(private val cityClient: WebClient, private val applicationName: String) : CityService {

    private val log = KotlinLogging.logger {}

    override suspend fun getCity(cityId: String): City {
        log.info { "Get city $cityId" }
        val requestId = deferContextual { mono { it.get<String>(requestKey) } }.awaitFirst()
        return cityClient.get()
            .uri("/cities/{id}", cityId)
            .header("X-Request-ID", requestId)
            .header("X-Source", applicationName)
            .retrieve()
            .awaitBody()
    }
}
